testsuite/tests/typing-modes/statefulness-visibility.ml

```diff
@@ -0,0 +1,552 @@
+[%%expect{|
+val default : 'a @ contended -> ('a @ immutable -> 'b) -> 'b = <fun>
+|}]
+
+let override : 'a @ contended -> ('a @ immutable uncontended -> 'b) -> 'b = fun x f -> f x
+[%%expect{|
+Line 1, characters 89-90:
+1 | let override : 'a @ contended -> ('a @ immutable uncontended -> 'b) -> 'b = fun x f -> f x
+                                                                                             ^
+Error: This value is "contended" but expected to be "uncontended".
+|}]
+
+let override : 'a @ contended -> ('a @ immutable shared -> 'b) -> 'b = fun x f -> f x
+[%%expect{|
+Line 1, characters 84-85:
+1 | let override : 'a @ contended -> ('a @ immutable shared -> 'b) -> 'b = fun x f -> f x
+                                                                                        ^
+Error: This value is "contended" but expected to be "shared".
+|}]
+
+(* [read] => [shared]. *)
+
+let default : 'a @ contended -> ('a @ read -> 'b) -> 'b = fun x f -> f x
+[%%expect{|
+Line 1, characters 71-72:
+1 | let default : 'a @ contended -> ('a @ read -> 'b) -> 'b = fun x f -> f x
+                                                                           ^
+Error: This value is "contended" but expected to be "shared".
+|}]
+
+let override : 'a @ contended -> ('a @ read uncontended -> 'b) -> 'b = fun x f -> f x
+[%%expect{|
+Line 1, characters 84-85:
+1 | let override : 'a @ contended -> ('a @ read uncontended -> 'b) -> 'b = fun x f -> f x
+                                                                                        ^
+Error: This value is "contended" but expected to be "uncontended".
+|}]
+
+let override : 'a @ contended -> ('a @ read contended -> 'b) -> 'b = fun x f -> f x
+[%%expect{|
+val override : 'a @ contended -> ('a @ contended read -> 'b) -> 'b = <fun>
+|}]
+
+(* [read_write] doesn't change the default. *)
+
+
+let fails : 'a @ contended -> ('a @ read_write uncontended -> 'b) -> 'b = fun x f -> f x
+[%%expect{|
+Line 1, characters 87-88:
+1 | let fails : 'a @ contended -> ('a @ read_write uncontended -> 'b) -> 'b = fun x f -> f x
+                                                                                           ^
+Error: This value is "contended" but expected to be "uncontended".
+|}]
+
+let fails : 'a @ contended -> ('a @ read_write shared -> 'b) -> 'b = fun x f -> f x
+[%%expect{|
+Line 1, characters 82-83:
+1 | let fails : 'a @ contended -> ('a @ read_write shared -> 'b) -> 'b = fun x f -> f x
+                                                                                      ^
+Error: This value is "contended" but expected to be "shared".
+|}]
+
+let succeeds : 'a @ contended -> ('a @ read_write contended -> 'b) -> 'b = fun x f -> f x
+[%%expect{|
+val succeeds : 'a @ contended -> ('a @ contended -> 'b) -> 'b = <fun>
+|}]
+
+(* Modalities. *)
+
+type 'a t1 = { y : 'a @@ immutable }
+
+let get : 'a @ contended -> 'a t1 = fun y -> {y}
+
+type 'a t2 = { z : 'a @@ shared }
```

Comment:
Should this be `read` rather than `shared`?

Reply:
Fixed.